docs/src/proposals/program-instruction-macro.md

```diff
@@ -0,0 +1,224 @@
+# Program Instruction Macro
+
+## Problem
+
+Currently, inspecting an on-chain transaction requires depending on a
+client-side, language-specific decoding library to parse the instruction.  If
+rpc methods could return decoded instruction details, these custom solutions
+would be unnecessary.
+
+We can deserialize instruction data using a program's Instruction enum, but
+decoding the account-key list into human-readable identifiers requires manual
+parsing. Our current Instruction enums have that account information, but only
+in variant docs.
+
+Similarly, we have instruction constructor functions that duplicate nearly all
+the information in the enum, but we can't generate that constructor from the
+enum definition because the list of account references is in code comments.
+
+Also, Instruction docs can vary between implementations, as there is no
+mechanism to ensure consistency.
+
+## Proposed Solution
+
+Move the data from code comments to attributes, such that the constructors
+can be generated, and include all the documentation from the enum definition.
+
+Here is an example of an Instruction enum using the new accounts format:
+
+```rust,ignore
+#[instructions(test_program::id())]
+pub enum TestInstruction {
+    /// Transfer lamports
+    #[accounts(
+        from_account(signer, writable, desc = "Funding account"),
```

Comment:
nit: `(from_account, signer, ...` instead of `from_account(`? Something to distinguish it from the function call syntax?  Also, `signer` instead of `SIGNER` looks like an identifier, not a constant.

Reply:
I used the `from_account(..` syntax because it is a standard attribute format, and a bit easier to parse in the macro. But I'll give your suggestion a go in implementation.
`signer/writer` updated to `SIGNER/WRITER`